File: js/02-add-two-numbers.js
function addTwoNumbers(l1, l2) {
    if(!l1 || !l2) return null

    let head = new ListNode(0, undefined)
    let current = head
    let carry = 0
    while(l1 || l2) {
        let x = l1 ? l1.val : 0
        let y = l2 ? l2.val : 0
        const sum = x + y + carry
        current.next = new ListNode(sum % 10, undefined)
        current = current.next
        carry = sum / 10 | 0
        if(l1) l1 = l1.next
        if(l2) l2 = l2.next
    }
    if(carry > 0) current.next = new ListNode(carry % 10, undefined)
    return head.next
};

File: go/02-add-two-numbers.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil || l2 == nil {
		return nil
	}

	head := &ListNode{Val: 0, Next: nil}
	carry := 0
	current := head

	for l1 != nil || l2 != nil {
		var x, y int
		if l1 == nil {
			x = 0
		} else {
			x = l1.Val
		}

		if l2 == nil {
			y = 0
		} else {
			y = l2.Val
		}

		current.Next = &ListNode{Val: (x + y + carry) % 10, Next: nil}
		current = current.Next
		carry = (x + y + carry) / 10

		if l1 != nil {
			l1 = l1.Next
		}

		if l2 != nil {
			l2 = l2.Next
		}
	}

	if carry > 0 {
		current.Next = &ListNode{Val: carry, Next: nil}
	}

	return head.Next
}
